### PR TITLE
[onnx][torch] Lower `onnx.grid_sampler` to the `torch` equivalents

### PR DESCRIPTION
**Pull request by afalkenberg1**
This is the lowering of gridsampler from onnx to torch using our prior implementation of AtenGridSamplerOp. 
Here are several checks for cornercases implemented. We may decide to have part of these checks in AtenGridSamplerOp instead of the onnx lowering portion.

### CODE REVIEW

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp [hunk 1]
```diff
@@ -92,6 +92,82 @@ void mlir::torch::onnx_c::populateDefaultDomainGtoP(
...
+        Value interpolationMode;
+        std::string padding;
+        Value paddingMode;
+        int64_t align;
+        Value alignCorners;
+
+        if (binder.tensorOperandAtIndex(input, 0) ||
+            binder.tensorOperandAtIndex(grid, 1) ||
+            binder.tensorResultType(resultType))
+          return failure();
```
Comment (rsuderman) [inline lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp:111]: Include a `rewriter.notifyMatchFailure()` just so we get better debug messages.

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp [hunk 2]
```diff
@@ -92,6 +92,82 @@ void mlir::torch::onnx_c::populateDefaultDomainGtoP(
...
+
+        auto inputTensorType = input.getType().cast<Torch::ValueTensorType>();
+        ArrayRef<int64_t> inputShape = inputTensorType.getSizes();
+        unsigned inputRank = inputShape.size();
```
Comment (rsuderman) [inline lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp:115]: Typically we prefer cstdint types, e.g. `int32_t`, `uint32_t` so we know the bitwidth. `unsigned` and `int` have undefined bit widths.

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp [hunk 3]
```diff
@@ -92,6 +92,82 @@ void mlir::torch::onnx_c::populateDefaultDomainGtoP(
...
+        unsigned gridRank = gridShape.size();
+
+        if (inputRank != 4) {
+          return rewriter.notifyMatchFailure(binder.op,
+                                             "only input rank 4 supported");
+        }
+        if (gridRank != 4) {
+          return rewriter.notifyMatchFailure(binder.op,
+                                             "only grid rank 4 supported");
+        }
```
Comment (rsuderman) [inline lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp:127]: You can drop the opening / closing brackets as this one just one line.

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp [hunk 4]
```diff
@@ -92,6 +92,82 @@ void mlir::torch::onnx_c::populateDefaultDomainGtoP(
...
+        if (binder.customOpNameStringAttr(mode, "mode", "bilinear"))
+          return rewriter.notifyMatchFailure(binder.op, "mode bind failure");
+        if (mode != "bilinear")
+          return rewriter.notifyMatchFailure(
+              binder.op, "currently only mode : bilinear supported");
+        interpolationMode = rewriter.create<Torch::ConstantIntOp>(
+            binder.getLoc(), rewriter.getType<Torch::IntType>(),
+            rewriter.getIntegerAttr(rewriter.getIntegerType(64), 0));
+
+        if (binder.customOpNameStringAttr(padding, "padding_mode", "zeros"))
```
Comment (rsuderman) [inline lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp:146]: You should do all of this validation before invoking a `rewriter.create` function.

**Review by rsuderman — CHANGES_REQUESTED** [review]

**Comment by afalkenberg1** [comment]
cool thanks

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp (hunk 1 above)
Comment (afalkenberg1) [reply lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp:111]: yes sounds better anyway

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp (hunk 2 above)
Comment (afalkenberg1) [reply lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp:115]: I agree with that notion, I usually do too but once in a while a unsinged slips into my writing.

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp (hunk 3 above)
Comment (afalkenberg1) [reply lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp:127]: cool yes have couple of those

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp (hunk 4 above)
Comment (afalkenberg1) [reply lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp:146]: yes that makes more sense

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp [hunk 5]
```diff
@@ -92,6 +92,77 @@ void mlir::torch::onnx_c::populateDefaultDomainGtoP(
...
         return success();
       });
+  patterns.onOp(
+      "GridSample", 20,
+      [](OpBinder binder, ConversionPatternRewriter &rewriter) {
+        Torch::ValueTensorType resultType;
+        Value input;
+        Value grid;
+        std::string mode;
+        Value interpolationMode;
```
Comment (rsuderman) [inline lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp:102]: Move these declarations to where they are used

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp [hunk 6]
```diff
@@ -92,6 +92,77 @@ void mlir::torch::onnx_c::populateDefaultDomainGtoP(
...
+          return rewriter.notifyMatchFailure(
+              binder.op, "currently only padding_mode : zeros supported");
+        if (binder.s64IntegerAttr(align, "align_corners", 0))
+          return rewriter.notifyMatchFailure(binder.op,
+                                             "align_corners bind failure");
+        if (align != 0)
+          return rewriter.notifyMatchFailure(
+              binder.op, "currently only align_corners : 0 supported");
+
+        interpolationMode = rewriter.create<Torch::ConstantIntOp>(
```
Comment (rsuderman) [inline lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp:151]: E.g. this line should be|
```
        Value interpolationMode = rewriter.create<Torch::ConstantIntOp>(
            binder.getLoc(), rewriter.getType<Torch::IntType>(),
            rewriter.getIntegerAttr(rewriter.getIntegerType(64), 0));
```

**Review by rsuderman — APPROVED** [review]

**Comment by rsuderman** [comment]
One more minor change then you can land

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp [hunk 7]
```diff
@@ -92,6 +92,77 @@ void mlir::torch::onnx_c::populateDefaultDomainGtoP(
...
+            binder.getLoc(), rewriter.getType<Torch::IntType>(),
+            rewriter.getIntegerAttr(rewriter.getIntegerType(64), 0));
+        paddingMode = rewriter.create<Torch::ConstantIntOp>(
```
Comment (rsuderman) [inline lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp:154]: And here

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp [hunk 8]
```diff
@@ -92,6 +92,77 @@ void mlir::torch::onnx_c::populateDefaultDomainGtoP(
...
+            binder.getLoc(), rewriter.getType<Torch::IntType>(),
+            rewriter.getIntegerAttr(rewriter.getIntegerType(64), 0));
+        alignCorners = rewriter.create<Torch::ConstantBoolOp>(
```
Comment (rsuderman) [inline lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp:157]: And here

Comment (afalkenberg1) [reply lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp]: ok did the changes as requested.

**Review by rsuderman — APPROVED** [review]

**Pull request merged**